### DISCUSSION TITLE
2FA w/ GitHub mobile option auto opt-in

### DISCUSSION THREAD

**Discussion by jonlabelle** [community/community#10861] [Mobile] [answered]
I didn't see an to option to opt out from sending auth codes via the GitHub mobile app (iOS) - apparently the only way to opt out is by uninstalling the app entirely :(

**Comment by eliperkins** [comment]
Hey @jonlabelle, thanks for this feedback. To help secure our customers, two-factor authentication via GitHub Mobile is enabled by default.

I'd love to hear a bit more about your expectations around disabling this, if you don't mind! I'd be happy to take your feedback back to our team to re-open this discussion.

- Where in the product would you expect to be able to opt-out of this authentication method?
- Once you've opted out, how would you expect to opt back in to re-enable GitHub Mobile 2FA?

**Reply by nihaals** [reply]
I think opting out on both the app (in settings somewhere) and on the website (maybe a button next to the device in [the device list](https://github.com/settings/two_factor_authentication/configure) and/or simply a button to disable GitHub Mobile 2FA) makes sense. When I tried to opt out I first looked on the website, then the settings page of the app, and then discovered there wasn't a way to do it.

Opting back in would probably make more sense to do from the device, depending on if the device is removed from the list in settings. If it's just showed with something like `disabled` next to it, then something similar to opting out would be intuitive.

Also, for the reason for opting out, I'm already satisfied with TOTP. I'd be fine with leaving GitHub Mobile enabled, but I still want my default to be TOTP instead of it requiring another button press and instantly sending a notification (although this could be seen as an advantage if someone else reaches the 2FA page). My end goal is more to have TOTP the default than to actually disable GitHub Mobile.

**Comment by JimRoepcke** [comment]
I configured an authenticator app. It's arrogant and reckless of GitHub to think it can change my security protocols unilaterally.

In fact, my settings still say they're configured to use the authenticator application, but GitHub is making me launch the mobile app on my device to enter a code shown on my browser instead.

<img width="691" alt="image" src="https://user-images.githubusercontent.com/39997/153352951-79d3e7ef-08fe-4842-b8ec-45dddaff66ea.png">

Like @nihaals I would be fine with leaving that method enabled if GitHub would respect my configured setting to use the authenticator application.

**Comment by FezVrasta** [comment]
Same issue, why is the mobile authentication the default option? I want to use my TOTP code that is automatically filled by my password manager.

**Comment by patrickwalton** [comment]
I want the opt out as well. My 2fa flow is ideal and GitHub mobile really throws it off.

**Comment by disrupted** [comment]
at least let us keep TOTP as the default

**Comment by mskelton** [comment]
I mean, GitHub is a platform for developers, people who generally understand a thing or two about computers and security.  Let us make security decisions that are right for us, don't be so heavy handed.

**Comment by humblehacker** [comment]
To add to the chorus: I use 1Password to enter my 2fa codes, and GitHub forcing me to use GitHub mobile as a first step adds friction to the login process. As a mobile developer, I commonly have many mobile devices on my desk, all set to do-not-disturb, so I'm not going to go hunting for a device to finish my GitHub login. Instead, I have to make the extra step of clicking through to be able to allow 1Password to complete the 2fa process.

Please refrain from making these sorts of decisions for us. These are different options, let us choose what works best for us. As it stands, my only recourse is to remove GitHub mobile from all of my devices to prevent this from a happening.

**Comment by eliperkins** [comment]
Thanks for all of the feedback here. We hear you on the friction that has been introduced as a part of the 2FA with GitHub Mobile ship. We're continuing to iterate on the feedback y'all have sent our way.

We've shipped new functionality on GitHub.com that will allow you to _always_ use a given 2FA method on a given machine. Always use your TOTP app when you're on your desktop browser? Perfect! Pick the TOTP option while logging into GitHub.com and choose 
"Use this method for future logins" to always use TOTP on that browser.

<img width="387" alt="Screen Shot 2022-03-08 at 11 53 37 AM" src="https://user-images.githubusercontent.com/1051453/157286732-573aa437-16ba-4fcc-abbf-287511ee4dcb.png">

Your feedback has been super valuable to us, and has helped us move forward to protect the accounts of not only all of you, but every GitHub developer. Please feel free to submit more feedback here in this repo to continue shaping the future of account security on GitHub!

**Reply by humblehacker** [reply]
This is a great step in the right direction. Thanks! But I'll reiterate that this should be a setting - since I'm using 1Password for 2FA, I'll never opt to use GitHub Mobile. So even though I can turn it off on an individual basis, I'll still be seeing this prompt from time-to-time on new devices. Please let me set it once and forget about it 🙏.

**Reply by humblehacker** [reply]
I'm returning here because I'm continuing to be frustrated by this feature. I don't want to disable GitHub Mobile 2FA "on a given machine", I want to disable it period. The current checkbox method isn't sufficient. 

First, just to be able to use the checkbox, I have to disable 1Password, as it will autofill the TOTP and GitHub will automatically continue as soon as the field is filled, without waiting for me to click "Verify", giving me no time to actually check the box.  So I have to disable 1Password, do the login dance, and re-enable 1Password. I shouldn't have to do that.

Second, the setting is stored in a cookie, which I commonly erase. So I'll have to do this little dance quite often. And even if I didn't delete my cookies, what if I use another browser? Or use a browser in a clean VM. I shouldn't have to do this same dance repeatedly.

You guys have clearly some internal requirement to drive people to use Github Mobile for 2FA, and you're using dark patterns to push us there. 

Dark patterns don't belong on Github. 

Please reconsider this implementation.

**Reply by drcdev** [reply]
> We've shipped new functionality on GitHub.com that will allow you to always use a given 2FA method on a given machine. Always use your TOTP app when you're on your desktop browser? Perfect! Pick the TOTP option while logging into GitHub.com and choose "Use this method for future logins" to always use TOTP on that browser.

This is still a huge issue, _especially_ on new devices. You should not be taking my use of the app as permission to use it for 2FA. I never set that up or agreed to it, much less gave permission to nudge me every time I log in from a new device or browser. There is no authentication in the app when I approve a login that way.

It should have a global preference to choose a default method at minimum. Until then, the app has to go!

**Reply by ChefAndy** [reply]
If you have an authentication setting that relies on a cookie then you don’t really have an authentication setting— you have a browser-based temporary preference. It has nothing to do with your device. How many web developers use GitHub? How many of them use one browser? How many of those browsers use default cookie handling? How often to people log into GitHub from a browser with their cookies intact?

**Reply by vishalnandagopal** [reply]
Please add a option to choose default 2FA method in a user's settings, or make the "Use this method for future logins" checkbox apply to all devices and not this specific device.

I clear cookies regularly and this makes GitHub revert back to the open mobile app option. I don't want to open the GitHub mobile app. Seems like the only way to disable the mobile app way of using 2FA is to uninstall the app, which I now have to do to avoid this annoying feature.

**Reply by ChefAndy** [reply]
@vishalnandagopal 

> I clear cookies regularly and this makes GitHub revert back to the open mobile app option.

Exactly. I don't often clear cookies in the browser I use GH in but I also stay logged in to GH, so I only need to log in when my cookies get blown away, so this setting is useless. Since I don't use the GH app for anything else, this is a super frustrating setup. I imagine between the two of us, we probably have the most common user stories for web logins.

**Reply by captainemre16** [reply]
yunusemrecftc16@gmail.com

**Reply by natevaughan** [reply]
> We've shipped new functionality on GitHub.com that will allow you to always use a given 2FA method on a given machine. Always use your TOTP app when you're on your desktop browser? Perfect! Pick the TOTP option while logging into GitHub.com and choose
"Use this method for future logins" to always use TOTP on that browser.

To be clear, you've implemented this as a _cookie_? So if someone clears their cookies, or uses a browser that automatically does so after closing a session, the default returns to the Github mobile app?

**Reply by Codel1417** [reply]
> > We've shipped new functionality on GitHub.com that will allow you to always use a given 2FA method on a given machine. Always use your TOTP app when you're on your desktop browser? Perfect! Pick the TOTP option while logging into GitHub.com and choose
> > "Use this method for future logins" to always use TOTP on that browser.
> 
> To be clear, you've implemented this as a _cookie_? So if someone clears their cookies, or uses a browser that automatically does so after closing a session, the default returns to the Github mobile app?

I would be more annoyed over some rouge dev selecting his own answer which is not what was asked or wanted

**Reply by iTrooz** [reply]
I'm thinking about uninstalling the mobile app because of this feature. I know you already heard this, but please reconsider the implementation

**Reply by moltar** [reply]
I'm with @humblehacker, this mobile 2fa is completely unexpected and breaks my workflow with 1Password 2FA setup. I often keep my phone away when working and it's extremely annoying to go fetch it. I later learned the "dark pattern", that there is a link at the bottom that allows me to switch the 2FA method, but still. It feels very intrusive for GH to change my 2FA flow without prior permission.

**Comment by guillaumevalverde** [comment]
But how do we opt out? if I lost my phone, i would like to disable it. However I dont see any option to do so (@eliperkins )

**Reply by eliperkins** [reply]
@guillaumevalverde You can revoke access for the [GitHub for iOS](https://github.com/settings/connections/applications/2cfa9b7a1b57de32dd0d) or [GitHub for Android](https://github.com/settings/connections/applications/3f8b8834a91f0caad392) OAuth apps from Settings > Applications > Authorized OAuth Apps and revoking the specific app from there. That will prevent sensitive interactions like 2FA from being able to be used by attackers who have physical access to your device.

GitHub Mobile 2FA uses biometric authentication to prevent attackers from approving authentication attempts, leveraging your device's built-in security hardware like the Secure Enclave on iOS or hardware-backed keystores on Android. If you were to lose your physical device, the attacker would still need your biometric data or device passcode to approve auth attempts using GitHub Mobile.

**Reply by guillaumevalverde** [reply]
Thanks for the info, I was looking for the term "device"  that s why i could not find it. Good point with the biometric topic.

**Reply by eliperkins** [reply]
> I was looking for the term "device" that s why i could not find it

Great point. I'll write up some docs on the GitHub Docs page to make this clearer 🙏

**Reply by jroose** [reply]
Why can't we opt-out from the 2FA menu?  That feels like a more natural place to put the opt-out.  Instead the 2FA menu only shows that it's enabled.  It's possible to manage every other form of MFA from that menu, and it's a bad user experience to need to navigate somewhere special to remove the automatic opt-in via a special menu.

**Reply by mselchow** [reply]
Additionally, removing the OAuth authorization from Settings does not stop the GitHub mobile 2FA prompts on login. I have signed out of the app, uninstalled it, revoked access, and it still asks me to authenticate via the app on login.

@eliperkins

**Reply by Joelsamijon** [reply]
⭐👍

**Reply by Rohit19060** [reply]
Revoking access is a must-have and very much accessible feature

**Comment by Ivan0414822** [comment]
> Hey @jonlabelle, thanks for this feedback. To help secure our customers, two-factor authentication via GitHub Mobile is enabled by default.
> 
> I'd love to hear a bit more about your expectations around disabling this, if you don't mind! I'd be happy to take your feedback back to our team to re-open this discussion.
> 
> - Where in the product would you expect to be able to opt-out of this authentication method?
> - Once you've opted out, how would you expect to opt back in to re-enable GitHub Mobile 2FA?

**Reply by Ivan0414822** [reply]
?????

**Reply by jonlabelle** [reply]
> Where in the product would you expect to be able to opt-out of this authentication method?

Under <https://github.com/settings/security>, **Access** -> ***Password and Authentication***
- GitHub Mobile
  - Then a simple **Enable/Disable** toggle

> Once you've opted out, how would you expect to opt back in to re-enable GitHub Mobile 2FA?

The same pathway as above.

**Reply by MagicalDrizzle** [reply]
> > Where in the product would you expect to be able to opt-out of this authentication method?
> 
> Under https://github.com/settings/security, **Access** -> _**Password and Authentication**_
> 
> * GitHub Mobile
>   
>   * Then a simple **Enable/Disable** toggle
> 
> > Once you've opted out, how would you expect to opt back in to re-enable GitHub Mobile 2FA?
> 
> The same pathway as above.

![Capture](https://user-images.githubusercontent.com/49761211/163718104-6fea68f3-ee64-48a0-b85f-3432fdd219aa.PNG)
It seems I can't even disable it at all?

**Comment by ChefAndy** [comment]
TLDR: I want SMS 2FA and I don't want it to stop being SMS 2FA until I tell it to stop being SMS 2FA. I really, _really_ don't want you or anybody else _changing my security settings for me._ Access to my repos is part of my basic daily logistical needs and someone else automatically switching that up with no notice is frankly, pretty scary. I can get SMS messages on my computer but I can't get GH Mobile pushes on my computer. By changing this, you're imposing the need, without notice, to have a specific device available to access my account. That is _broken_.

SMS 2FA worked fine until I wiped my phone I'd guess. Best I can figure is that GH Mobile automatically logging in using the credentials stored on the icloud backup was enough to zap my 2FA settings. I took zero direct action related to my (paid for almost a decade, mind you) GH account at all. I mean— wth.

I re-set up 2FA in my account specifically using SMS and not the app. I can't find any way of removing GH mobile and no settings in the app seem to be remotely appropriate. So I should delete the app? What if I can’t get back in? I didn't see any way to deliberately choose SMS when presented with the app challenge screen, so if I lose access to the app, even if I have SMS set up, I'm looking at recovery code path, right? The whole reason I switched to SMS in the first place was being burned by recovery codes not working after a device reset killed my DUO link. I lost access to my GitHub account for nearly 48 hours waiting for support to get around to it. The GH authentication system already lost my trust, and just bagged my mitigation strategy. I've had this phone number for 21 years— let me make decisions about what makes sense for my account. 

More importantly, I’ve been leaving my phone in my locker at work because phones kill my and many other people’s attention. I have a diagnosed attention deficit— widely recognized to be covered by the ADA— so consider that before strong-arming people into unlocking and interacting with their phones at work.

I saw you asking above where people would expect to find this opt-out— just use your existing wireframes and triple underline the copy because it just doesn’t behave like the interface implies. 

PS: This might be my misconception, but MS seems allergic lightweight, simple features focused on efficiently solving people’s problems rather than increasing adoption of MS ProblemSolution Professional™— _now with extra extra extra features!!!_ This is exactly the sort of thing so many people were worried about when MS bought GH. Please prove me abjectly, embarrassingly wrong with your savvy UX and Product Design decisions focused on user empowerment and solving people's problems with _the smallest possible overhead._

**Comment by bviktor** [comment]
Worst misfeature ever.

I DELETED the darn app a long time ago, but GitHub still wants to authenticate me using that.

I want to disable this piece of sh*t, so I go to "reconfigure 2FA", I go through all the trouble to add the GH account to MS Authenticator, download the recovery codes AGAIN, and guess what:

![image](https://user-images.githubusercontent.com/5798552/163880644-99090d45-ba6b-4a7e-a46d-d7a3d5e4b165.png)

It LEAVES the friggin' app enabled! Are you mental?

So I try to disable 2FA (now I really feel how much you "improved" my security by shoving this stupid feature down my throat), but guess what: I belong to an organization that "requires" it.

So I LEAVE the organization. Disable 2FA, re-enable 2FA without EVER touching GitHub Mobile, and guess what??

![image](https://user-images.githubusercontent.com/5798552/163880644-99090d45-ba6b-4a7e-a46d-d7a3d5e4b165.png)

It still uses the god damned GH app!

So then I reinstall GH Mobile, and after minutes of looking for "deregistering", I eventually just "log out", and finally this PoS site doesn't want to use GH Mobile to log in.

THANKS A LOT GITHUB, IT'S SUCH A NICE FEATURE TO KNOW BETTER THAN ME HOW I WANT TO LOG IN.

Also thanks for ensuring I'll NEVER EVER install your darned app on my phone again. What a great way to waste my time. Way to go, above and beyond.

**Reply by MagicalDrizzle** [reply]
hey at least you solve it by just logging out of the app
I have done that and it stills beg me to use it lmao
someone said it's a dark pattern, and I can't say I disagree...

**Comment by rungrotArunon** [comment]
Thanks you bro

**Reply by Codel1417** [reply]
It's been 2 more months and you still can't opt out

**Comment by humblehacker** [comment]
Since this has been marked as "Answered", maybe the folks at GitHub are no longer paying attention to it? Anyone want to start a new discussion?

**Reply by effleurager** [reply]
Let's hope they don't close https://github.com/orgs/github-community/discussions/18753 claiming it's a duplicate.

**Comment by sammcj** [comment]
The end of the year is fast approaching and this is still an issue.

Any idea what's going on @eliperkins?

**Comment by ashtonian** [comment]
This is pretty half baked and annoying ux.

I don't want to use the mobile app for 2fa.
The "Use this method for future logins" checkbox implies that this is a global setting but it seems to be a cookie setting which is useless as I often don't re login to GitHub in my home browsers.
The [settings](https://github.com/settings/security) -> Password and authentication -> Two-factor authentication section has a "Primary two-factor method" option which would seem to indicate preference, and it does.. until you install the mobile app.
Simply uninstalling the mobile doesn't work, you have to revoke its access in the settings for this to get fixed.

**Comment by apjyotirmay** [comment]
Uninstalled GitHub mobile, because this was annoying af. Please be better.

**Reply by iTrooz** [reply]
Agreed, I use https://github.com/git-touch/git-touch instead of the official app because of this

**Reply by sammcj** [reply]
That's a flutter "app" though - you're probably better off just using the browser.

**Reply by apjyotirmay** [reply]
Right. I'm simply using the website on my desktop and mobile. It's overall a better experience compared to app + website, sadly.

**Comment by sammcj** [comment]
I'm pretty surprised this hasn't been given priority to be fixed, it's a major turn off for the mobile app.

As usual Github/Microsoft seem more focused on throwing features at the platform than fixing fundamental issues.

**Comment by joeldeteves** [comment]
Please get rid of the garbage feature. Listen to your userbase - nobody wants it!

**Comment by ghost** [comment]
> I didn't see an to option to opt out from sending auth codes via the GitHub mobile app (android) - apparently the only way to opt out is by uninstalling the app entirely :(

**Comment by eliperkins** [comment, marked as answer]
Thanks for your feedback, y'all! Today, we've shipped [a new feature that allows for setting a preferred option for two-factor authentication](https://github.blog/changelog/2023-02-22-preferred-2fa-methods-and-settings-improvements/). This will allow you to set other two-factor authentication methods to be used before another option. [Go to Settings > Password and Authentication on GitHub.com](https://github.com/settings/security) in your browser to set a different preferred two-factor option.

<img width="772" alt="Screenshot 2023-02-22 at 4 46 23 PM" src="https://user-images.githubusercontent.com/1051453/220766920-fc0cacc4-5539-4afc-9940-83acbd17efee.png">

Additionally, [GitHub Mobile sessions can be revoked within Settings](https://github.com/settings/sessions?#mobile_sessions) as well:

<img width="793" alt="Screenshot 2023-02-22 at 4 49 54 PM" src="https://user-images.githubusercontent.com/1051453/220767378-9bdca298-994d-4cb8-8b7c-22724b16e2a1.png">

**Reply by drcdev** [reply]
This is a huge improvement! I was able to re-install the GitHub Mobile app again without disrupting my desktop flow. Thank you for listening :)

**Reply by Codel1417** [reply]
Ok, but can you disable GitHub mobile from self registering as a 2fa option?

**Reply by ClaraLeigh** [reply]
Whats the point of having a hardware 2fa if you can just bypass it with github mobile? It's really disappointing that you're forced to have this enabled

**Reply by kotatsuyaki** [reply]
I uninstalled the GitHub mobile app from all my mobile devices.
Please, provide an option to disable it as a 2FA factor.

**Reply by csutcliff** [reply]
> I uninstalled the GitHub mobile app from all my mobile devices.
> Please, provide an option to disable it as a 2FA factor.

They did, 2 months ago

**Reply by kotatsuyaki** [reply]
> > I uninstalled the GitHub mobile app from all my mobile devices.
> > Please, provide an option to disable it as a 2FA factor.
> 
> They did, 2 months ago

As far as I can tell, I'm able to set TOTP as the preferred one, but I can't completely disable GitHub mobile as a 2FA factor. Am I missing something on the settings page?

**Reply by Codel1417** [reply]
> > > I uninstalled the GitHub mobile app from all my mobile devices.
> > > Please, provide an option to disable it as a 2FA factor.
> > 
> > 
> > They did, 2 months ago
> 
> As far as I can tell, I'm able to set TOTP as the preferred one, but I can't completely disable GitHub mobile as a 2FA factor. Am I missing something on the settings page?

Nope, GitHub falsely marked this discussion as solved without addressing the issue. I worry for the platform if they are  marked issues as solved without fixing the issue.

**Reply by Wladefant** [reply]
@amrehman @candyho @eliperkins this discussion is now resolved and can be closed

**Reply by Codel1417** [reply]
Can you login to GitHub mobile without it auto enabling APP 2FA?

**Comment by Spixmacaw** [comment]
Can you login to GitHub mobile without it auto enabling APP 2FA?
